File: p605e/can_place_flowers.js
/**
 * @param {number[]} flowerbed
 * @param {number} n
 * @return {boolean}
 */
var canPlaceFlowers = function (flowerbed, n) {
  flowerbed = [0].concat(flowerbed).concat([0]);
  for (let i = 0; i < flowerbed.length - 1; ++i) {
    if (
      flowerbed[i - 1] === 0 &&
      flowerbed[i] === 0 &&
      flowerbed[i + 1] === 0
    ) {
      flowerbed[i] = 1;
      if (--n == 0) return true;
    }
  }
  return n <= 0;
};
// TEST
[
  [[0], 1, true],
  [[0, 0], 1, true],
  [[1, 0, 1], 1, false],
  [[1, 0, 0, 0, 1], 1, true],
  [[1, 0, 0, 0, 1], 2, false],
].forEach(([flowerbed, n, expected]) => {
  const actual = canPlaceFlowers(flowerbed, n);
  console.log('Can place', n, 'flowers in', flowerbed, '->', actual);
  console.assert(actual === expected);
});

File: p605e/can_place_flowers.py
class Solution:
    def canPlaceFlowers(self, flowerbed: list[int], n: int) -> bool:
        flowerbed = [0] + flowerbed + [0]
        for i in range(1, len(flowerbed) - 1):
            if flowerbed[i - 1] == flowerbed[i] == flowerbed[i + 1] == 0:
                n, flowerbed[i] = n - 1, 1
                if n == 0:
                    return True
        return n <= 0


# TESTS
for flowerbed, n, expected in [
    ([0], 1, True),
    ([0, 0], 1, True),
    ([1, 0, 1], 1, False),
    ([1, 0, 0, 0, 1], 1, True),
    ([1, 0, 0, 0, 1], 2, False),
]:
    sol = Solution()
    actual = sol.canPlaceFlowers(flowerbed, n)
    print(n, "flowers can be planted in", flowerbed, "->", actual)
    assert actual == expected
